File: module_ssh/src/main/java/com/xiaoyv/ssh/terminal/model/TerminalModel.java
package com.xiaoyv.ssh.terminal.model;

import com.blankj.utilcode.util.ThreadUtils;
import com.trilead.ssh2.Connection;
import com.trilead.ssh2.Session;
import com.xiaoyv.busines.config.SshLoginType;
import com.xiaoyv.busines.exception.RxException;
import com.xiaoyv.busines.room.entity.SshEntity;
import com.xiaoyv.ssh.terminal.contract.TerminalContract;

import java.io.File;

import io.reactivex.rxjava3.core.Observable;

/**
 * TerminalModel
 *
 * @author why
 * @since 2020/12/06
 **/
public class TerminalModel implements TerminalContract.Model {
    private Connection connection;

    @Override
    public Observable<Session> p2mConnectSsh(SshEntity sshEntity) {
        return Observable.create(emitter -> {
            try {
                int port = Integer.parseInt(sshEntity.port);
                connection = new Connection(sshEntity.ip, port);
                connection.connect((sHost, sPort, sEntType, bytes) -> true, 5000, 5000);
                boolean authenticate = false;
                switch (sshEntity.authType) {
                    case SshLoginType.TYPE_NONE:
                        authenticate = connection.authenticateWithNone(sshEntity.account);
                        break;
                    case SshLoginType.TYPE_PASSWORD:
                        authenticate = connection.authenticateWithPassword(sshEntity.account, sshEntity.password);
                        break;
                    case SshLoginType.TYPE_PUBLIC_KEY:
                        authenticate = connection.authenticateWithPublicKey(sshEntity.account, new File(""), sshEntity.password);
                        break;
                    default:
                        break;
                }
                // 连接失败
                if (!authenticate) {
                    emitter.onError(new RxException("SSH 认证失败"));
                    emitter.onComplete();
                    return;
                }
                Session session = connection.openSession();
                session.requestPTY("vt100");
                session.startShell();

                emitter.onNext(session);
                emitter.onComplete();
            } catch (Exception e) {
                emitter.onError(e);
                emitter.onComplete();
            }
        });
    }

    @Override
    public void p2mClose() {
        if (connection != null) {
            ThreadUtils.getCachedPool().execute(() -> {
                connection.close();
                connection = null;
            });
        }
    }
}

File: module_ssh/src/main/java/com/xiaoyv/ssh/terminal/TerminalModel.kt
@file:Suppress("MemberVisibilityCanBePrivate")

package com.xiaoyv.ssh.terminal

import androidx.annotation.WorkerThread
import com.trilead.ssh2.Connection
import com.trilead.ssh2.ServerHostKeyVerifier
import com.trilead.ssh2.Session
import com.xiaoyv.blueprint.base.rxjava.event.RxEvent
import com.xiaoyv.blueprint.rxbus.RxBus
import com.xiaoyv.busines.config.SshLoginType
import com.xiaoyv.busines.room.entity.SshEntity
import com.xiaoyv.busines.rx.RxEventTag
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.schedulers.Schedulers
import java.io.File
import java.io.IOException

/**
 * TerminalModel
 *
 * @author why
 * @since 2020/12/06
 */
class TerminalModel : TerminalContract.Model {

    private var sshConnection: Connection? = null
    private var sshSession: Session? = null

    override fun p2mConnectSsh(sshEntity: SshEntity): Observable<Session> {
        return Observable.create {
            val connection = createConnect(sshEntity)

            // 会话
            val session = connection.openSession().apply {
                requestPTY("xterm")
                startShell()
            }

            // 监听
            connection.addConnectionMonitor {
                RxBus.getDefault().post(RxEvent(), RxEventTag.EVENT_SSH_DISCONNECT)
            }

            // 缓存对象
            sshSession = session
            sshConnection = connection

            it.onNext(session)
            it.onComplete()
        }
    }

    override fun p2mReleaseSession() {
        Observable.create<Boolean> {
            closeConnectionQuietly()

            it.onNext(true)
            it.onComplete()
        }.observeOn(Schedulers.io())
            .subscribeOn(Schedulers.io())
            .subscribe()
    }

    override fun p2mSendHeartbeatPackets() {
        sshConnection?.sendIgnorePacket()
    }

    /**
     * 静默关闭连接
     */
    private fun closeConnectionQuietly() {
        runCatching {
            sshSession?.close()
            sshSession = null
        }

        runCatching {
            sshConnection?.close()
            sshConnection = null
        }
    }

    companion object {

        /**
         * 连接前验证
         */
        private val serverHostKeyVerifier = ServerHostKeyVerifier { sHost, sPort, sEntType, bytes ->
            true
        }

        @JvmStatic
        @WorkerThread
        @Throws(IOException::class)
        fun createConnect(sshEntity: SshEntity): Connection {
            val port = sshEntity.port.toIntOrNull() ?: 22
            val connection = Connection(sshEntity.ip, port)
            try {
                connection.connect(serverHostKeyVerifier, 5000, 5000)

                var authenticate = false

                // 授权类型
                when (sshEntity.authType) {
                    SshLoginType.TYPE_NONE -> {
                        authenticate = connection.authenticateWithNone(sshEntity.account)
                    }
                    SshLoginType.TYPE_PASSWORD -> {
                        authenticate =
                            connection.authenticateWithPassword(
                                sshEntity.account,
                                sshEntity.password
                            )
                    }
                    SshLoginType.TYPE_PUBLIC_KEY -> {
                        authenticate =
                            connection.authenticateWithPublicKey(
                                sshEntity.account,
                                File(""),
                                sshEntity.password
                            )
                    }
                }

                if (authenticate) {
                    return connection
                }
                throw IOException("无法认证 SSH")
            } catch (e: Exception) {
                throw IOException("无法连接 SSH", e)
            }
        }
    }
}